services/archiver/archiver.go

```diff
@@ -0,0 +1,306 @@
+// Copyright 2020 The Gitea Authors.
+// All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package archiver
+
+import (
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"code.gitea.io/gitea/modules/base"
+	"code.gitea.io/gitea/modules/context"
+	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+
+	"github.com/unknwon/com"
+)
+
+// ArchiveRequest defines the parameters of an archive request, which notably
+// includes the specific repository being archived as well as the commit, the
+// name by which it was requested, and the kind of archive being requested.
+// This is entirely opaque to external entities, though, and mostly used as a
+// handle elsewhere.
+type ArchiveRequest struct {
+	uri             string
+	repo            *git.Repository
+	refName         string
+	ext             string
+	archivePath     string
+	archiveType     git.ArchiveType
+	archiveComplete bool
+	commit          *git.Commit
+	cchan           chan struct{}
+}
+
+var archiveInProgress []*ArchiveRequest
+var archiveMutex sync.Mutex
+
+// SHA1 hashes will only go up to 40 characters, but SHA256 hashes will go all
+// the way to 64.
+var shaRegex = regexp.MustCompile(`^[0-9a-f]{4,64}$`)
+
+// These facilitate testing, by allowing the unit tests to control (to some extent)
+// the goroutine used for processing the queue.
+var archiveQueueMutex *sync.Mutex
+var archiveQueueStartCond *sync.Cond
+var archiveQueueReleaseCond *sync.Cond
+
+// GetArchivePath returns the path from which we can serve this archive.
+func (aReq *ArchiveRequest) GetArchivePath() string {
+	return aReq.archivePath
+}
+
+// GetArchiveName returns the name of the caller, based on the ref used by the
+// caller to create this request.
+func (aReq *ArchiveRequest) GetArchiveName() string {
+	return aReq.refName + aReq.ext
+}
+
+// IsComplete returns the completion status of this request.
+func (aReq *ArchiveRequest) IsComplete() bool {
+	return aReq.archiveComplete
+}
+
+// WaitForCompletion will wait for this request to complete, with no timeout.
+// It returns whether the archive was actually completed, as the channel could
+// have also been closed due to an error.
+func (aReq *ArchiveRequest) WaitForCompletion() bool {
+	<-aReq.cchan
+	return aReq.IsComplete()
+}
+
+// TimedWaitForCompletion will wait for this request to complete, with timeout
+// happening after the specified Duration.  It returns whether the archive is
+// now complete and whether we hit the timeout or not.  The latter may not be
+// useful if the request is complete.
+func (aReq *ArchiveRequest) TimedWaitForCompletion(dur time.Duration) (bool, bool) {
+	select {
+	case <-time.After(dur):
+		return aReq.IsComplete(), true
+	case <-aReq.cchan:
+		return aReq.IsComplete(), false
+	}
```

Comment:
Even if we expect `dur` to be short, we should run this in a "with shutdown" context as well.